patches/server/0725-Custom-Advancement-Triggers.patch

```diff
@@ -0,0 +1,192 @@
+@@ -0,0 +1,52 @@
++package io.papermc.paper.advancements;
++
++import com.google.gson.JsonObject;
++import io.papermc.paper.advancements.CraftCustomAdvancementTrigger.CustomAdvancementTriggerInstance;
++import net.minecraft.advancements.critereon.EntityPredicate;
++import net.minecraft.advancements.critereon.SimpleCriterionTrigger;
++import net.minecraft.advancements.critereon.AbstractCriterionTriggerInstance;
++import net.minecraft.advancements.critereon.DeserializationContext;
++import net.minecraft.resources.ResourceLocation;
++import org.bukkit.NamespacedKey;
++import org.bukkit.craftbukkit.entity.CraftPlayer;
++import org.bukkit.craftbukkit.util.CraftNamespacedKey;
++import org.bukkit.entity.Player;
++
++import java.util.function.Predicate;
++
++public class CraftCustomAdvancementTrigger<T extends CustomAdvancementTrigger<D>, D> extends SimpleCriterionTrigger<CustomAdvancementTriggerInstance<D>> {
++
++    private final ResourceLocation key;
++    private final T trigger;
++
++    public CraftCustomAdvancementTrigger(NamespacedKey key, T trigger) {
++        this.key = CraftNamespacedKey.toMinecraft(key);
++        this.trigger = trigger;
++    }
++
++    @Override
++    public ResourceLocation getId() {
++        return key;
++    }
++
++    public void trigger(Player player, Predicate<D> instancePredicate) {
++        this.trigger(((CraftPlayer) player).getHandle(), customAdvancementTriggerInstance -> {
++            return instancePredicate.test(customAdvancementTriggerInstance.data);
++        });
++    }
++
++    @Override
++    protected CustomAdvancementTriggerInstance<D> createInstance(JsonObject conditionsObject, EntityPredicate.Composite playerPredicate, DeserializationContext predicateDeserializer) {
++        return new CustomAdvancementTriggerInstance<>(key, playerPredicate, trigger.createInstance(conditionsObject));
++    }
++
++    public static class CustomAdvancementTriggerInstance<D> extends AbstractCriterionTriggerInstance {
++
++        private final D data;
++
++        public CustomAdvancementTriggerInstance(ResourceLocation minecraftkey, EntityPredicate.Composite criterionconditionentity_b, D data) {
++            super(minecraftkey, criterionconditionentity_b);
++            this.data = data;
++        }
++    }
++}
+diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
+index cfd43069ee2b6f79afb12e10d223f6bf75100034..daff65e6db8fd1319586fa6043b0e926d3c67fb0 100644
+--- a/src/main/java/net/minecraft/server/Main.java
++++ b/src/main/java/net/minecraft/server/Main.java
+@@ -184,6 +184,7 @@ public class Main {
+             }
+             // CraftBukkit end
+             DataPackConfig datapackconfiguration1 = MinecraftServer.configurePackRepository(resourcepackrepository, datapackconfiguration == null ? DataPackConfig.DEFAULT : datapackconfiguration, flag);
++            /* Paper start - move loading datapacks until after plugins
```

Comment:
Why?

Reply:
Because, any datapack that uses the custom trigger that the plugin adds will fail to load because the plugin hasn't registered it's own yet.

Reply:
Add that in the comment, please

Reply:
Done